contracts/Staking.sol

```diff
@@ -0,0 +1,191 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.6.0;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@nomiclabs/buidler/console.sol";
+
+contract Staking {
+    // timestamp for the epoch 1
+    // everything before that is considered epoch 0 which won't have a reward but allows for the initial stake
+    uint256 public epoch1Start;
+
+    // duration of each epoch
+    uint256 epochDuration;
+
+    mapping(address => mapping(address => uint256)) private balances;
+
+    struct Pool {
+        uint256 size;
+        bool set;
+    }
+
+    // for each token, we store the total pool size
+    mapping(address => mapping(uint256 => Pool)) private poolSize;
+
+    struct Checkpoint {
+        uint256 epochId;
+        uint256 balance;
+    }
+
+    // balanceCheckpoints[user][token][]
+    mapping(address => mapping(address => Checkpoint[])) private balanceCheckpoints;
+
+    /*
+     *
+     */
+    constructor (uint256 _epoch1Start, uint256 _epochDuration) public {
+        epoch1Start = _epoch1Start;
+        epochDuration = _epochDuration;
+    }
+
+    /*
+     * Stores `amount` of `tokenAddress` tokens for the `user` into the vault
+     */
+    function deposit(address tokenAddress, uint256 amount) public {
+        require(amount > 0, "Staking: Amount must be > 0");
+
+        IERC20 token = IERC20(tokenAddress);
+        uint256 allowance = token.allowance(msg.sender, address(this));
+        require(allowance >= amount, "Staking: Token allowance too small");
+
+        balances[msg.sender][tokenAddress] += amount;
+
+        token.transferFrom(msg.sender, address(this), amount);
+
+        // epoch logic
+        uint256 currentEpoch = getCurrentEpoch();
+        uint256 nextEpoch = currentEpoch + 1;
+
+        // update the next epoch pool size
+        Pool storage pNextEpoch = poolSize[tokenAddress][nextEpoch];
+        pNextEpoch.size = token.balanceOf(address(this));
+        pNextEpoch.set = true;
+
+        Checkpoint[] storage checkpoints = balanceCheckpoints[msg.sender][tokenAddress];
+        if ((checkpoints.length == 0) || (checkpoints[checkpoints.length - 1].epochId != nextEpoch)){
+            checkpoints.push(Checkpoint(nextEpoch, balances[msg.sender][tokenAddress]));
```

Comment:
again. before pusjing the nextepoch checkpoint we need to check that checkpoints[checkpoints.length - 1].epochId != currentEpoch because otherwise if you get a withdraw just a moment after a deposit the note written in the withdraw is not valid anymore (the user has a balance but for the next epoch not for the current)

Reply:
Good catch! Fixed this in the latest version, added a test case for it and also fixed a related edge case.